# Pandas

A Pandas series is a one-dimensional array-like object that can hold many data types, such as numbers or strings. One of the main differences between Pandas Series and NumPy ndarrays is that you can assign an index label to each element in the Pandas Series. In other words, you can name the indices of your Pandas Series anything you want. Another big difference between Pandas Series and NumPy ndarrays is that Pandas Series can hold data of different data types.
Let's start by importing Pandas into Python. It has become a convention to import Pandas as <code>pd</code>, therefore, you can import Pandas by typing the following command in your Jupyter notebook:

## Creating Pandas Series

Let's begin by creating a Pandas Series. You can create Pandas Series by using the command <code>pd.Series(data, index)</code>, where index is a list of index labels. Let's use a Pandas Series to store a grocery list. We will use the food items as index labels and the quantity we need to buy of each item as our data.

We import Pandas as <code>pd</code> into Python 

In [1]:
import pandas as pd

We create a Pandas Series that stores a grocery list

In [2]:
groceries = pd.Series(data = [30, 6, 'Yes', 'No'], index = ['eggs', 'apples', 'milk', 'bread'])

We display the Groceries Pandas Series

In [3]:
groceries

eggs       30
apples      6
milk      Yes
bread      No
dtype: object

We see that Pandas Series are displayed with the indices in the first column and the data in the second column. Notice that the data is not indexed 0 to 3 but rather it is indexed with the names of the food we put in, namely eggs, apples, etc… Also notice that the data in our Pandas Series has both integers and strings.
Just like NumPy ndarrays, Pandas Series have attributes that allows us to get information from the series in an easy way. Let's see some of them:

We print some information about Groceries

In [4]:
print('Groceries has shape:', groceries.shape)
print('Groceries has dimension:', groceries.ndim)
print('Groceries has a total of:', groceries.size, 'elements')

Groceries has shape: (4,)
Groceries has dimension: 1
Groceries has a total of: 4 elements


We can also print the index labels and the data of the Pandas Series separately. This is useful if you don't happen to know what the index labels of the Pandas Series are

We print the index and data of Groceries

In [5]:
print('The data in Groceries is:', groceries.values)
print('The index of Groceries is:', groceries.index)

The data in Groceries is: [30 6 'Yes' 'No']
The index of Groceries is: Index(['eggs', 'apples', 'milk', 'bread'], dtype='object')


If you are dealing with a very large Pandas Series and if you are not sure whether an index label exists, you can check by using the in command

We check whether bananas is a food item (an index) in Groceries

In [6]:
x = 'bananas' in groceries

We check whether bread is a food item (an index) in Groceries

In [7]:
y = 'bread' in groceries

We print the results

In [8]:
print('Is bananas an index label in Groceries:', x)
print('Is bread an index label in Groceries:', y)

Is bananas an index label in Groceries: False
Is bread an index label in Groceries: True


## Accessing and Deleting Elements in Pandas Series

Now let's look at how we can access or modify elements in a Pandas Series. One great advantage of Pandas Series is that it allows us to access data in many different ways. Elements can be accessed using index labels or numerical indices inside square brackets, [ ], similar to how we access elements in NumPy ndarrays. Since we can use numerical indices, we can use both positive and negative integers to access data from the beginning or from the end of the Series, respectively. Since we can access elements in various ways, in order to remove any ambiguity to whether we are referring to an index label or numerical index, Pandas Series have two attributes, <code>.loc</code> and <code>.iloc</code> to explicitly state what we mean. The attribute .loc stands for location and it is used to explicitly state that we are using a labeled index. Similarly, the attribute <code> .iloc</code> stands for integer location and it is used to explicitly state that we are using a numerical index. Let's see some examples:

We access elements in Groceries using index labels:

We use a single index label

In [9]:
print('How many eggs do we need to buy:', groceries['eggs'])

How many eggs do we need to buy: 30


we can access multiple index labels

In [10]:
print('Do we need milk and bread:\n', groceries[['milk', 'bread']]) 

Do we need milk and bread:
 milk     Yes
bread     No
dtype: object


we use loc to access multiple index labels

In [11]:
print('How many eggs and apples do we need to buy:\n', groceries.loc[['eggs', 'apples']]) 

How many eggs and apples do we need to buy:
 eggs      30
apples     6
dtype: object


We access elements in Groceries using numerical indices:

we use multiple numerical indices

In [12]:
print('How many eggs and apples do we need to buy:\n',  groceries[[0, 1]]) 

How many eggs and apples do we need to buy:
 eggs      30
apples     6
dtype: object


We use a negative numerical index

In [13]:
print('Do we need bread:\n', groceries[[-1]])

Do we need bread:
 bread    No
dtype: object


We use a single numerical index

In [14]:
print('How many eggs do we need to buy:', groceries[0])

How many eggs do we need to buy: 30


we use iloc to access multiple numerical indices

In [15]:
print('Do we need milk and bread:\n', groceries.iloc[[2, 3]]) 

Do we need milk and bread:
 milk     Yes
bread     No
dtype: object


Pandas Series are also mutable like NumPy ndarrays, which means we can change the elements of a Pandas Series after it has been created. For example, let's change the number of eggs we need to buy from our grocery list

We display the original grocery list

In [16]:
print('Original Grocery List:\n', groceries)

Original Grocery List:
 eggs       30
apples      6
milk      Yes
bread      No
dtype: object


We change the number of eggs to 2

In [17]:
groceries['eggs'] = 2

We display the changed grocery list

In [18]:
print('Modified Grocery List:\n', groceries)

Modified Grocery List:
 eggs        2
apples      6
milk      Yes
bread      No
dtype: object


We can also delete items from a Pandas Series by using the <code>.drop()</code> method. The Series<code>.drop(label) </code> method removes the given label from the given Series. We should note that the <code>Series.drop(label)</code> method drops elements from the Series out of place, meaning that it doesn't change the original Series being modified. Let's see how this works:

We display the original grocery list

In [19]:
print('Original Grocery List:\n', groceries)

Original Grocery List:
 eggs        2
apples      6
milk      Yes
bread      No
dtype: object


We remove apples from our grocery list. The drop function removes elements out of place

In [20]:
print('We remove apples (out of place):\n', groceries.drop('apples'))

We remove apples (out of place):
 eggs       2
milk     Yes
bread     No
dtype: object


When we remove elements out of place the original Series remains intact. To see this
we display our grocery list again

In [21]:
print('Grocery List after removing apples out of place:\n', groceries)

Grocery List after removing apples out of place:
 eggs        2
apples      6
milk      Yes
bread      No
dtype: object


We can delete items from a Pandas Series in place by setting the keyword inplace to True in the .drop() method. Let's see an example:

We display the original grocery list

In [22]:
print('Original Grocery List:\n', groceries)

Original Grocery List:
 eggs        2
apples      6
milk      Yes
bread      No
dtype: object


We remove apples from our grocery list in place by setting the inplace keyword to True

In [23]:
groceries.drop('apples', inplace = True)

When we remove elements in place the original Series its modified. To see this
we display our grocery list again

In [24]:
print('Grocery List after removing apples in place:\n', groceries)

Grocery List after removing apples in place:
 eggs       2
milk     Yes
bread     No
dtype: object


## Arithmetic Operations on Pandas Series

Just like with NumPy ndarrays, we can perform element-wise arithmetic operations on Pandas Series. In this lesson we will look at arithmetic operations between Pandas Series and single numbers. Let's create a new Pandas Series that will hold a grocery list of just fruits.

We create a Pandas Series that stores a grocery list of just fruits

In [25]:
fruits= pd.Series(data = [10,6,3], index = ['apples', 'oranges', 'bananas'])

We display the fruits Pandas Series

In [26]:
fruits

apples     10
oranges     6
bananas     3
dtype: int64

We can now modify the data in fruits by performing basic arithmetic operations. Let's see some examples

We print fruits for reference

In [27]:
print('Original grocery list of fruits:\n ', fruits)

Original grocery list of fruits:
  apples     10
oranges     6
bananas     3
dtype: int64


 We perform basic element-wise operations using arithmetic symbols

In [28]:
print('fruits + 2:\n', fruits + 2) # We add 2 to each item in fruits

fruits + 2:
 apples     12
oranges     8
bananas     5
dtype: int64


In [29]:
print('fruits - 2:\n', fruits - 2) # We subtract 2 to each item in fruits

fruits - 2:
 apples     8
oranges    4
bananas    1
dtype: int64


In [30]:
print('fruits * 2:\n', fruits * 2) # We multiply each item in fruits by 2 

fruits * 2:
 apples     20
oranges    12
bananas     6
dtype: int64


In [31]:
print('fruits / 2:\n', fruits / 2) # We divide each item in fruits by 2

fruits / 2:
 apples     5.0
oranges    3.0
bananas    1.5
dtype: float64


You can also apply mathematical functions from NumPy, such as<code>sqrt(x)</code>, to all elements of a Pandas Series.

We import NumPy as np to be able to use the mathematical functions

In [32]:
import numpy as np

We print fruits for reference

In [33]:
print('Original grocery list of fruits:\n', fruits)

Original grocery list of fruits:
 apples     10
oranges     6
bananas     3
dtype: int64


We apply different mathematical functions to all elements of fruits

In [34]:
print('EXP(X) = \n', np.exp(fruits))

EXP(X) = 
 apples     22026.465795
oranges      403.428793
bananas       20.085537
dtype: float64


In [35]:
print('SQRT(X) = \n', np.sqrt(fruits))

SQRT(X) = 
 apples     3.162278
oranges    2.449490
bananas    1.732051
dtype: float64


In [36]:
print('POW(X,2) =\n',np.power(fruits,2)) # We raise all elements of fruits to the power of 2

POW(X,2) =
 apples     100
oranges     36
bananas      9
dtype: int64


Pandas also allows us to only apply arithmetic operations on selected items in our fruits grocery list. Let's see some examples

We print fruits for reference

In [37]:
print('Original grocery list of fruits: \n', fruits)

Original grocery list of fruits: 
 apples     10
oranges     6
bananas     3
dtype: int64


In [38]:
print('Amount of bananas + 2 = ', fruits['bananas'] + 2)

Amount of bananas + 2 =  5


In [39]:
print('Amount of apples - 2 = ', fruits.iloc[0]-2)

Amount of apples - 2 =  8


In [40]:
print('We double the amount of apples and oranges:\n', fruits[['apples', 'oranges']] * 2)

We double the amount of apples and oranges:
 apples     20
oranges    12
dtype: int64


In [41]:
print('We half the amount of apples and oranges:\n', fruits.loc[['apples', 'oranges']] / 2)

We half the amount of apples and oranges:
 apples     5.0
oranges    3.0
dtype: float64


You can also apply arithmetic operations on Pandas Series of mixed data type provided that the arithmetic operation is defined for all data types in the Series, otherwise you will get an error. Let's see what happens when we multiply our grocery list by 2

We multiply our grocery list by 2

In [42]:
groceries * 2

eggs          4
milk     YesYes
bread      NoNo
dtype: object

As we can see, in this case, since we multiplied by 2, Pandas doubles the data of each item including the strings. Pandas can do this because the multiplication operation <code> * </code> is defined both for numbers and strings. If you were to apply an operation that was valid for numbers but not strings, say for instance, <code>/</code> you will get an error. So when you have mixed data types in your Pandas Series make sure the arithmetic operations are valid on all the data types of your elements.

## Manipulate a Series (Exercise)

In [43]:
# Create a Pandas Series that contains the distance of some planets from the Sun.
# Use the name of the planets as the index to your Pandas Series, and the distance
# from the Sun as your data. The distance from the Sun is in units of 10^6 km

distance_from_sun = [149.6, 1433.5, 227.9, 108.2, 778.6]

planets = ['Earth','Saturn', 'Mars','Venus', 'Jupiter']


In [44]:
# Create a Pandas Series using the above data, with the name of the planets as
# the index and the distance from the Sun as your data.
dist_planets = pd.Series(data = distance_from_sun, index = planets)

In [45]:
# Calculate the number of minutes it takes sunlight to reach each planet. You can
# do this by dividing the distance from the Sun for each planet by the speed of light.
# Since in the data above the distance from the Sun is in units of 10^6 km, you can
# use a value for the speed of light of c = 18, since light travels 18 x 10^6 km/minute.
time_light = dist_planets / 18

In [46]:
# Use Boolean indexing to select only those planets for which sunlight takes less
# than 40 minutes to reach them.
close_planets = time_light[time_light < 40]

## Creating Pandas DataFrames

We create a dictionary of Pandas Series

In [47]:
items = {'Bob' : pd.Series(data = [245, 25, 55], index = ['bike', 'pants', 'watch']),
        'Alice' : pd.Series(data = [40, 110, 500, 45], index = ['book','glasses', 'bike','pants'])}

We print the type of items to see that it is a dictionary

In [48]:
print(type(items))

<class 'dict'>


Now that we have a dictionary, we are ready to create a DataFrame by passing it to the pd.DataFrame() function. We will create a DataFrame that could represent the shopping carts of various users, in this case we have only two users, Alice and Bob.

We create a Pandas DataFrame by passing it a dictionary of Pandas Series

In [49]:
shopping_carts = pd.DataFrame(items)

We display the DataFrame

In [50]:
shopping_carts

Bob  Alice
bike     245.0  500.0
book       NaN   40.0
glasses    NaN  110.0
pants     25.0   45.0
watch     55.0    NaN

There are several things to notice here that are worth pointing out. We see that DataFrames are displayed in tabular form, much like an Excel spreadsheet, with the labels of rows and columns in bold. Also notice that the row labels of the DataFrame are built from the union of the index labels of the two Pandas Series we used to construct the dictionary. And the column labels of the DataFrame are taken from the keys of the dictionary. Another thing to notice is that the columns are arranged alphabetically and not in the order given in the dictionary. We will see later that this won't happen when we load data into a DataFrame from a data file. The last thing we want to point out is that we see some NaN values appear in the DataFrame. NaN stands for Not a Number, and is Pandas way of indicating that it doesn't have a value for that particular row and column index. For example, if we look at the column of Alice, we see that it has NaN in the watch index. You can see why this is the case by looking at the dictionary we created at the beginning. We clearly see that the dictionary has no item for Alice labeled watches. So whenever a DataFrame is created, if a particular column doesn't have values for a particular row index, Pandas will put a NaN value there. If we were to feed this data into a machine learning algorithm we will have to remove these NaN values first. In a later lesson we will learn how to deal with NaN values and clean our data. For now, we will leave these values in our DataFrame.
In the above example we created a Pandas DataFrame from a dictionary of Pandas Series that had clearly defined indexes. If we don't provide index labels to the Pandas Series, Pandas will use numerical row indexes when it creates the DataFrame. Let's see an example:

We create a dictionary of Pandas Series without indexes

In [51]:
data = {'Bob' : pd.Series([245, 25, 55]),
        'Alice' : pd.Series([40, 110, 500, 45])}

We create a DataFrame

In [52]:
df = pd.DataFrame(data)

We display the DataFrame

In [53]:
df

Bob  Alice
0  245.0     40
1   25.0    110
2   55.0    500
3    NaN     45

We can see that Pandas indexes the rows of the DataFrame starting from 0, just like NumPy indexes ndarrays.
Now, just like with Pandas Series we can also extract information from DataFrames using attributes. Let's print some information from our <code>shopping_carts</code> DataFrame

In [54]:
print('shopping_carts has shape:', shopping_carts.shape)
print('shopping_carts has dimension:', shopping_carts.ndim)
print('shopping_carts has a total of:', shopping_carts.size, 'elements')
print()
print('The data in shopping_carts is:\n', shopping_carts.values)
print()
print('The row index in shopping_carts is:', shopping_carts.index)
print()
print('The column index in shopping_carts is:', shopping_carts.columns)

shopping_carts has shape: (5, 2)
shopping_carts has dimension: 2
shopping_carts has a total of: 10 elements

The data in shopping_carts is:
 [[245. 500.]
 [ nan  40.]
 [ nan 110.]
 [ 25.  45.]
 [ 55.  nan]]

The row index in shopping_carts is: Index(['bike', 'book', 'glasses', 'pants', 'watch'], dtype='object')

The column index in shopping_carts is: Index(['Bob', 'Alice'], dtype='object')


When creating the <code>shopping_carts</code> DataFrame we passed the entire dictionary to the <code>pd.DataFrame()</code> function. However, there might be cases when you are only interested in a subset of the data. Pandas allows us to select which data we want to put into our DataFrame by means of the keywords <code>columns</code> and <code>index</code>. Let's see some examples:

We Create a DataFrame that only has Bob's data

In [55]:
bob_shopping_cart = pd.DataFrame(items, columns=['Bob'])

We display bob_shopping_cart

In [56]:
bob_shopping_cart

Bob
bike   245
pants   25
watch   55

We Create a DataFrame that only has selected items for both Alice and Bob

In [57]:
sel_shopping_cart = pd.DataFrame(items, index = ['pants', 'book'])

We display sel_shopping_cart

In [58]:
sel_shopping_cart

Bob  Alice
pants  25.0     45
book    NaN     40

We Create a DataFrame that only has selected items for Alice

In [59]:
alice_sel_shopping_cart = pd.DataFrame(items, index = ['glasses', 'bike'], columns = ['Alice'])

We display alice_sel_shopping_cart

In [60]:
alice_sel_shopping_cart

Alice
glasses    110
bike       500

We create a dictionary of lists (arrays)

In [61]:
data = {'Integers' : [1,2,3],
        'Floats' : [4.5, 8.2, 9.6]}

We create a DataFrame 

In [62]:
df = pd.DataFrame(data)

We display the DataFrame

In [63]:
df

Integers  Floats
0         1     4.5
1         2     8.2
2         3     9.6

Notice that since the <code>data</code> dictionary we created doesn't have label indices, Pandas automatically uses numerical row indexes when it creates the DataFrame. We can however, put labels to the row index by using the <code>index</code> keyword in the <code>pd.DataFrame()</code> function. Let's see an example

 We create a list of Python dictionaries

In [64]:
items2 = [{'bikes': 20, 'pants': 30, 'watches': 35}, 
          {'watches': 10, 'glasses': 50, 'bikes': 15, 'pants':5}]

We create a DataFrame

In [65]:
store_items = pd.DataFrame(items2)

We display the DataFrame

In [66]:
store_items

bikes  glasses  pants  watches
0     20      NaN     30       35
1     15     50.0      5       10

Again, notice that since the <code>items2</code> dictionary we created doesn't have label indices, Pandas automatically uses numerical row indexes when it creates the DataFrame. As before, we can put labels to the row index by using the <code>index</code> keyword in the <code>pd.DataFrame()</code> function. Let's assume we are going to use this DataFrame to hold the number of items a particular store has in stock. So, we will label the row indices as store 1 and store 2.

We create a list of Python dictionaries

In [67]:
items2 = [{'bikes': 20, 'pants': 30, 'watches': 35}, 
          {'watches': 10, 'glasses': 50, 'bikes': 15, 'pants':5}]

We create a DataFrame  and provide the row index

In [68]:
store_items = pd.DataFrame(items2, index = ['store 1', 'store 2'])

We display the DataFrame

In [69]:
store_items

bikes  glasses  pants  watches
store 1     20      NaN     30       35
store 2     15     50.0      5       10

We can access elements in Pandas DataFrames in many different ways. In general, we can access rows, columns, or individual elements of the DataFrame by using the row and column labels. We will use the same <code>store_items</code> DataFrame created in the previous lesson. Let's see some examples:

In [70]:
print('How many bikes are in each store:\n', store_items[('bikes')])

How many bikes are in each store:
 store 1    20
store 2    15
Name: bikes, dtype: int64


In [71]:
print('How many bikes and pants are in each store:\n', store_items[['bikes', 'pants']])

How many bikes and pants are in each store:
          bikes  pants
store 1     20     30
store 2     15      5


In [72]:
print('What items are in Store 1:\n', store_items.loc[['store 1']])

What items are in Store 1:
          bikes  glasses  pants  watches
store 1     20      NaN     30       35


In [73]:
print('How many bikes are in Store 2:', store_items['bikes']['store 2'])

How many bikes are in Store 2: 15


We add a new column named shirts to our store_items DataFrame indicating the number of shirts in stock at each store. We will put 15 shirts in store 1 and 2 shirts in store 2

In [74]:
store_items['shirts'] = [15, 2]

We display the modified DataFrame

In [75]:
store_items

bikes  glasses  pants  watches  shirts
store 1     20      NaN     30       35      15
store 2     15     50.0      5       10       2

We can see that when we add a new column, the new column is added at the end of our DataFrame.

We can also add new columns to our DataFrame by using arithmetic operations between other columns in our DataFrame. Let's see an example:

We make a new column called suits by adding the number of shirts and pants

In [76]:
store_items['suits'] = store_items['pants']+store_items['shirts']

We display the modified DataFrame

In [77]:
store_items

bikes  glasses  pants  watches  shirts  suits
store 1     20      NaN     30       35      15     45
store 2     15     50.0      5       10       2      7

Suppose now, that you opened a new store and you need to add the number of items in stock of that new store into your DataFrame. We can do this by adding a new row to the <code>store_items</code> Dataframe. To add rows to our DataFrame we first have to create a new Dataframe and then append it to the original DataFrame. Let's see how this works

 We create a dictionary from a list of Python dictionaries that will number of items at the new store

In [78]:
new_items = [{'bikes': 20, 'pants': 30, 'watches': 35, 'glasses': 4}]

We create new DataFrame with the new_items and provide and index labeled store 3

In [79]:
new_store = pd.DataFrame(new_items, index = ['store 3'])

In [80]:
new_store

bikes  glasses  pants  watches
store 3     20        4     30       35

We now add this row to our <code>store_items</code> DataFrame by using the <code>.append()</code> method.

We append store 3 to our store_items DataFrame

In [81]:
store_items = store_items.append(new_store)

C:\Users\HP Pavilion\Anaconda3\lib\site-packages\pandas\core\frame.py:6692: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  sort=sort)


We display the modified DataFrame

In [82]:
store_items

bikes  glasses  pants  shirts  suits  watches
store 1     20      NaN     30    15.0   45.0       35
store 2     15     50.0      5     2.0    7.0       10
store 3     20      4.0     30     NaN    NaN       35

Notice that by appending a new row to the DataFrame, the columns have been put in alphabetical order.

We can also add new columns of our DataFrame by using only data from particular rows in particular columns. For example, suppose that you want to stock stores 2 and 3 with new watches and you want the quantity of the new watches to be the same as the watches already in stock for those stores. Let's see how we can do this

 We add a new column using data from particular rows in the watches column

In [83]:
store_items['new watches'] = store_items['watches'][1:]

We display the modified DataFrame

In [84]:
store_items

bikes  glasses  pants  shirts  suits  watches  new watches
store 1     20      NaN     30    15.0   45.0       35          NaN
store 2     15     50.0      5     2.0    7.0       10         10.0
store 3     20      4.0     30     NaN    NaN       35         35.0

It is also possible, to insert new columns into the DataFrames anywhere we want. The <code>dataframe.insert(loc,label,data)</code> method allows us to insert a new column in the <code>dataframe</code> at location <code>loc</code>, with the given column <code>label</code>, and given <code>data</code>. Let's add new column named <code>shoes</code> right before the **suits** column. Since suits has numerical index value 4 then we will use this value as **loc**. Let's see how this works:

 We insert a new column with label shoes right before the column with numerical index 4

In [85]:
store_items.insert(4, 'shoes', [8,5,0])

 we display the modified DataFrame

In [86]:
store_items

bikes  glasses  pants  shirts  shoes  suits  watches  new watches
store 1     20      NaN     30    15.0      8   45.0       35          NaN
store 2     15     50.0      5     2.0      5    7.0       10         10.0
store 3     20      4.0     30     NaN      0    NaN       35         35.0

Just as we can add rows and columns we can also delete them. To delete rows and columns from our DataFrame we will use the <code>.pop()</code> and <code>.drop()</code> methods. The <code>.pop()</code> method only allows us to delete columns, while the <code>.drop()</code> method can be used to delete both rows and columns by use of the axis keyword. Let's see some examples

We remove the new watches column

In [87]:
store_items.pop('new watches')

store 1     NaN
store 2    10.0
store 3    35.0
Name: new watches, dtype: float64

we display the modified DataFrame

In [88]:
store_items

bikes  glasses  pants  shirts  shoes  suits  watches
store 1     20      NaN     30    15.0      8   45.0       35
store 2     15     50.0      5     2.0      5    7.0       10
store 3     20      4.0     30     NaN      0    NaN       35

We remove the store 2 and store 1 rows

In [89]:
store_items = store_items.drop(['store 2', 'store 1'], axis = 0)

we display the modified DataFrame

In [90]:
store_items

bikes  glasses  pants  shirts  shoes  suits  watches
store 3     20      4.0     30     NaN      0    NaN       35

Sometimes we might need to change the row and column labels. Let's change the **bikes** column label to hats using the **.rename()** method

We change the column label bikes to hats

In [91]:
store_items = store_items.rename(columns = {'bikes': 'hats'})

we display the modified DataFrame

In [92]:
store_items

hats  glasses  pants  shirts  shoes  suits  watches
store 3    20      4.0     30     NaN      0    NaN       35

Now let's change the row label using the <code>.rename()</code> method again.

In [93]:
store_items = store_items.rename(index = {'store 3': 'last store'})

we display the modified DataFrame

In [94]:
store_items

hats  glasses  pants  shirts  shoes  suits  watches
last store    20      4.0     30     NaN      0    NaN       35

You can also change the index to be one of the columns in the DataFrame.

 We change the row index to be the data in the pants column.

In [95]:
store_items = store_items.set_index('pants')

we display the modified DataFrame

In [96]:
store_items

hats  glasses  shirts  shoes  suits  watches
pants                                              
30       20      4.0     NaN      0    NaN       35

## Dealing with NaN

As mentioned earlier, before we can begin training our learning algorithms with large datasets, we usually need to clean the data first. This means we need to have a method for detecting and correcting errors in our data. While any given dataset can have many types of bad data, such as outliers or incorrect values, the type of bad data we encounter almost always is missing values. As we saw earlier, Pandas assigns NaN values to missing data. In this lesson we will learn how to detect and deal with NaN values.

We will begin by creating a DataFrame with some NaN values in it

We create a list of Python dictionaries

In [97]:
items2 = [{'bikes': 20, 'pants': 30, 'watches': 35, 'shirts': 15, 'shoes': 8, 'suits':45},
         {'watches': 10, 'glasses': 50, 'bikes': 15, 'pants': 5, 'shirts': 2, 'shoes':5, 'suits': 7},
         {'bikes': 20, 'pants': 30, 'watches': 35, 'glasses': 4, 'shoes': 10}]

We create a DataFrame  and provide the row index

In [98]:
store_items = pd.DataFrame(items2, index = ['store 1', 'store 2','store 3'])

We display the DataFrame

In [99]:
store_items

bikes  glasses  pants  shirts  shoes  suits  watches
store 1     20      NaN     30    15.0      8   45.0       35
store 2     15     50.0      5     2.0      5    7.0       10
store 3     20      4.0     30     NaN     10    NaN       35

We can clearly see that the DataFrame we created has 3 NaN values: one in store 1 and two in store 3. However, in cases where we load very large datasets into a DataFrame, possibly with millions of items, the number of NaN values is not easily visualized. For these cases, we can use a combination of methods to count the number of NaN values in our data. The following example combines the <code>.isnull()</code> and the <code>sum()</code> methods to count the number of NaN values in our DataFrame

We count the number of NaN values in store_items

In [100]:
x = store_items.isnull().sum().sum()

We print x

In [101]:
print('Number of NaN values in our DataFrame:', x)

Number of NaN values in our DataFrame: 3


In the above example, the <code>.isnull()</code> method returns a *Boolean* DataFrame of the same size as <code>store_items</code> and indicates with True the elements that have NaN values and with False the elements that are not. Let's see an example:

In [102]:
store_items.isnull()

bikes  glasses  pants  shirts  shoes  suits  watches
store 1  False     True  False   False  False  False    False
store 2  False    False  False   False  False  False    False
store 3  False    False  False    True  False   True    False

In Pandas, logical True values have numerical value 1 and logical False values have numerical value 0. Therefore, we can count the number of NaN values by counting the number of logical True values. In order to count the total number of logical True values we use the <code>.sum()</code> method twice. We have to use it twice because the first sum returns a Pandas Series with the sums of logical True values along columns, as we see below:

In [103]:
store_items.isnull().sum()

bikes      0
glasses    1
pants      0
shirts     1
shoes      0
suits      1
watches    0
dtype: int64

The second sum will then add up the 1s in the above Pandas Series.

Instead of counting the number of NaN values we can also do the opposite, we can count the number of non-NaN values. We can do this by using the <code>.count()</code> method as shown below:

‌We print the number of non-NaN values in our DataFrame

In [104]:
print('Number of non-NaN values in the columns of our DataFrame:\n', store_items.count())

Number of non-NaN values in the columns of our DataFrame:
 bikes      3
glasses    2
pants      3
shirts     2
shoes      3
suits      2
watches    3
dtype: int64


Now that we learned how to know if our dataset has any <code>NaN</code> values in it, the next step is to decide what to do with them. In general we have two options, we can either delete or replace the NaN values. In the following examples we will show you how to do both.

We will start by learning how to eliminate rows or columns from our DataFrame that contain any NaN values. The .dropna(axis) method eliminates any rows with NaN values when axis = 0 is used and will eliminate any columns with NaN values when axis = 1 is used. Let's see some examples

We drop any rows with NaN values

In [105]:
store_items.dropna(axis = 0)

bikes  glasses  pants  shirts  shoes  suits  watches
store 2     15     50.0      5     2.0      5    7.0       10

We drop any columns with NaN values

In [106]:
store_items.dropna(axis = 1)

bikes  pants  shoes  watches
store 1     20     30      8       35
store 2     15      5      5       10
store 3     20     30     10       35

Notice that the .dropna() method eliminates (drops) the rows or columns with NaN values out of place. This means that the original DataFrame is not modified. You can always remove the desired rows or columns in place by setting the keyword <code>inplace = True</code> inside the <code>dropna()</code> function.

Now, instead of eliminating NaN values, we can replace them with suitable values. We could choose for example to replace all NaN values with the value 0. We can do this by using the <code>.fillna()</code> method as shown below.

We replace all NaN values with 0

In [107]:
store_items.fillna(0)

bikes  glasses  pants  shirts  shoes  suits  watches
store 1     20      0.0     30    15.0      8   45.0       35
store 2     15     50.0      5     2.0      5    7.0       10
store 3     20      4.0     30     0.0     10    0.0       35

We can also use the <code>.fillna()</code> method to replace NaN values with previous values in the DataFrame, this is known as forward filling. When replacing <code>NaN</code> values with forward filling, we can use previous values taken from columns or rows. The <code>.fillna(method = 'ffill', axis)</code> will use the forward filling (ffill) method to replace NaN values using the previous known value along the given axis. Let's see some examples:

In [108]:
store_items.fillna(method = 'ffill', axis = 0)

bikes  glasses  pants  shirts  shoes  suits  watches
store 1     20      NaN     30    15.0      8   45.0       35
store 2     15     50.0      5     2.0      5    7.0       10
store 3     20      4.0     30     2.0     10    7.0       35

Notice that the two <code>NaN</code> values in store 3 have been replaced with previous values in their columns. However, notice that the <code>NaN</code> value in store 1 didn't get replaced. That's because there are no previous values in this column, since the NaN value is the first value in that column. However, if we do forward fill using the previous row values, this won't happen. Let's take a look:

 We replace NaN values with the previous value in the row

In [109]:
store_items.fillna(method = 'ffill', axis = 1)

bikes  glasses  pants  shirts  shoes  suits  watches
store 1   20.0     20.0   30.0    15.0    8.0   45.0     35.0
store 2   15.0     50.0    5.0     2.0    5.0    7.0     10.0
store 3   20.0      4.0   30.0    30.0   10.0   10.0     35.0

Similarly, you can choose to replace the <code>NaN</code> values with the values that go after them in the DataFrame, this is known as backward filling. The <code>.fillna(method = 'backfill', axis)</code> will use the backward filling (backfill) method to replace NaN values using the next known value along the given axis. Just like with forward filling we can choose to use row or column values. Let's see some examples:

We replace NaN values with the next value in the column

In [110]:
store_items.fillna(method = 'backfill', axis = 0)

bikes  glasses  pants  shirts  shoes  suits  watches
store 1     20     50.0     30    15.0      8   45.0       35
store 2     15     50.0      5     2.0      5    7.0       10
store 3     20      4.0     30     NaN     10    NaN       35

Notice that the NaN value in store 1 has been replaced with the next value in its column. However, notice that the two NaN values in store 3 didn't get replaced. That's because there are no next values in these columns, since these NaN values are the last values in those columns. However, if we do backward fill using the next row values, this won't happen. Let's take a look:

We replace NaN values with the next value in the row

In [111]:
store_items.fillna(method = 'backfill', axis = 1)

bikes  glasses  pants  shirts  shoes  suits  watches
store 1   20.0     30.0   30.0    15.0    8.0   45.0     35.0
store 2   15.0     50.0    5.0     2.0    5.0    7.0     10.0
store 3   20.0      4.0   30.0    10.0   10.0   35.0     35.0

 Notice that the <code>.fillna()</code> method replaces (fills) the NaN values out of place. This means that the original DataFrame is not modified. You can always replace the NaN values in place by setting the keyword <code>inplace = True</code> inside the <code>fillna()</code> function.

We can also choose to replace NaN values by using different interpolation methods. For example, the <code>.interpolate(method = 'linear', axis)</code> method will use linear interpolation to replace NaN values using the values along the given axis. Let's see some examples:

We replace NaN values by using linear interpolation using column values

In [112]:
store_items.interpolate(method = 'linear', axis = 0)

bikes  glasses  pants  shirts  shoes  suits  watches
store 1     20      NaN     30    15.0      8   45.0       35
store 2     15     50.0      5     2.0      5    7.0       10
store 3     20      4.0     30     2.0     10    7.0       35

Notice that the two NaN values in store 3 have been replaced with linear interpolated values. However, notice that the NaN value in store 1 didn't get replaced. That's because the NaN value is the first value in that column, and since there is no data before it, the interpolation function can't calculate a value. Now, let's interpolate using row values instead:

 We replace NaN values by using linear interpolation using row values

In [113]:
store_items.interpolate(method = 'linear', axis = 1)

bikes  glasses  pants  shirts  shoes  suits  watches
store 1   20.0     25.0   30.0    15.0    8.0   45.0     35.0
store 2   15.0     50.0    5.0     2.0    5.0    7.0     10.0
store 3   20.0      4.0   30.0    20.0   10.0   22.5     35.0

Just as with the other methods we saw, the .interpolate() method replaces NaN values out of place.

## Manipulate a DataFrame

In [114]:
# Since we will be working with ratings, we will set the precision of our 
# dataframes to one decimal place.
pd.set_option('precision', 1)

# Create a Pandas DataFrame that contains the ratings some users have given to a
# series of books. The ratings given are in the range from 1 to 5, with 5 being
# the best score. The names of the books, the authors, and the ratings of each user
# are given below:

books = pd.Series(data = ['Great Expectations', 'Of Mice and Men', 'Romeo and Juliet', 'The Time Machine', 'Alice in Wonderland' ])
authors = pd.Series(data = ['Charles Dickens', 'John Steinbeck', 'William Shakespeare', ' H. G. Wells', 'Lewis Carroll' ])

user_1 = pd.Series(data = [3.2, np.nan ,2.5])
user_2 = pd.Series(data = [5., 1.3, 4.0, 3.8])
user_3 = pd.Series(data = [2.0, 2.3, np.nan, 4])
user_4 = pd.Series(data = [4, 3.5, 4, 5, 4.2])

# Users that have np.nan values means that the user has not yet rated that book.
# Use the data above to create a Pandas DataFrame that has the following column
# labels: 'Author', 'Book Title', 'User 1', 'User 2', 'User 3', 'User 4'. Let Pandas
# automatically assign numerical row indices to the DataFrame. 

# Create a dictionary with the data given above
dat = {'Book Title' : books,
      'Author' : authors,
      'User 1' : user_1,
      'User 2' : user_2,
      'User 3' : user_3,
      'User 4' : user_4}

# Use the dictionary to create a Pandas DataFrame
book_ratings = pd.DataFrame(dat)

# If you created the dictionary correctly you should have a Pandas DataFrame
# that has column labels: 'Author', 'Book Title', 'User 1', 'User 2', 'User 3',
# 'User 4' and row indices 0 through 4.

# Now replace all the NaN values in your DataFrame with the average rating in
# each column. Replace the NaN values in place. HINT: you can use the fillna()
# function with the keyword inplace = True, to do this. Write your code below:
# book_ratings.fillna(book_ratings.mean(),inplace = True)

## Loading Data into a Pandas DataFrame

In machine learning you will most likely use databases from many sources to train your learning algorithms. Pandas allows us to load databases of different formats into DataFrames. One of the most popular data formats used to store databases is csv. CSV stands for Comma Separated Values and offers a simple format to store data. We can load CSV files into Pandas DataFrames using the <code>pd.read_csv()</code> function. Let's load Google stock data into a Pandas DataFrame. The GOOG.csv file contains Google stock data from 8/19/2004 till 10/13/2017 taken from Yahoo Finance.

We load Google stock data in a DataFrame

In [116]:
Google_stock = pd.read_csv('./GOOG.csv')

We print some information about Google_stock

In [117]:
print('Google_stock is of type:', type(Google_stock))
print('Google_stock has shape:', Google_stock.shape)

Google_stock is of type: <class 'pandas.core.frame.DataFrame'>
Google_stock has shape: (253, 7)


We see that we have loaded the GOOG.csv file into a Pandas DataFrame and it consists of 253 rows and 7 columns. Now let's look at the stock data

In [118]:
Google_stock

Date    Open    High     Low   Close  Adj Close   Volume
0    2019-01-31  1103.0  1117.3  1095.4  1116.4     1116.4  1538300
1    2019-02-01  1112.4  1125.0  1104.9  1110.8     1110.8  1462200
2    2019-02-04  1112.7  1132.8  1109.0  1132.8     1132.8  2576500
3    2019-02-05  1124.8  1146.8  1117.2  1146.0     1146.0  3552200
4    2019-02-06  1139.6  1147.0  1112.8  1115.2     1115.2  2105600
5    2019-02-07  1104.2  1104.8  1086.0  1098.7     1098.7  2044800
6    2019-02-08  1087.0  1098.9  1086.6  1095.1     1095.1  1075800
7    2019-02-11  1096.9  1105.9  1092.9  1095.0     1095.0  1065200
8    2019-02-12  1106.8  1125.3  1105.8  1121.4     1121.4  1609100
9    2019-02-13  1125.0  1134.7  1118.5  1120.2     1120.2  1049800
10   2019-02-14  1118.1  1128.2  1110.4  1121.7     1121.7   947600
11   2019-02-15  1130.1  1131.7  1110.7  1113.7     1113.7  1449800
12   2019-02-19  1110.0  1121.9  1110.0  1118.6     1118.6  1046400
13   2019-02-20  1120.0  1123.4  1105.3  1113.8     1113.8  1087800
14   2019-02-21  1110.8  1111.9  1092.5  1097.0     1097.0  1415100
15   2019-02-22  1100.9  1111.2  1095.6  1110.4     1110.4  1049500
16   2019-02-25  1116.0  1118.5  1107.3  1109.4     1109.4  1413100
17   2019-02-26  1105.8  1119.5  1099.9  1115.1     1115.1  1471300
18   2019-02-27  1106.9  1118.0  1101.0  1116.1     1116.1   968400
19   2019-02-28  1111.3  1127.7  1111.0  1119.9     1119.9  1542500
20   2019-03-01  1124.9  1143.0  1124.8  1141.0     1141.0  1450300
21   2019-03-04  1147.0  1158.3  1130.7  1147.8     1147.8  1446000
22   2019-03-05  1150.1  1169.6  1146.2  1162.0     1162.0  1443200
23   2019-03-06  1162.5  1167.6  1155.5  1157.9     1157.9  1099300
24   2019-03-07  1155.7  1156.8  1134.9  1143.3     1143.3  1166600
25   2019-03-08  1126.7  1147.1  1123.3  1142.3     1142.3  1212400
26   2019-03-11  1144.4  1176.2  1144.4  1175.8     1175.8  1719200
27   2019-03-12  1178.3  1200.0  1178.3  1193.2     1193.2  2013100
28   2019-03-13  1200.6  1200.9  1191.9  1193.3     1193.3  1435900
29   2019-03-14  1194.5  1197.9  1184.5  1185.6     1185.6  1172800
..          ...     ...     ...     ...     ...        ...      ...
223  2019-12-18  1356.6  1360.5  1351.0  1352.6     1352.6  1522600
224  2019-12-19  1351.8  1358.1  1349.0  1356.0     1356.0  1469900
225  2019-12-20  1363.3  1363.6  1349.0  1349.6     1349.6  3315000
226  2019-12-23  1355.9  1359.8  1346.5  1348.8     1348.8   883100
227  2019-12-24  1348.5  1350.3  1342.8  1343.6     1343.6   347500
228  2019-12-26  1346.2  1361.3  1344.5  1360.4     1360.4   667500
229  2019-12-27  1363.0  1364.5  1349.3  1351.9     1351.9  1038400
230  2019-12-30  1350.0  1353.0  1334.0  1336.1     1336.1  1050900
231  2019-12-31  1330.1  1338.0  1329.1  1337.0     1337.0   961800
232  2020-01-02  1341.6  1368.1  1341.6  1367.4     1367.4  1406600
233  2020-01-03  1347.9  1372.5  1345.5  1360.7     1360.7  1186400
234  2020-01-06  1350.0  1396.5  1350.0  1394.2     1394.2  1732300
235  2020-01-07  1397.9  1403.0  1390.4  1393.3     1393.3  1502700
236  2020-01-08  1392.1  1411.6  1390.8  1404.3     1404.3  1528000
237  2020-01-09  1420.6  1427.3  1410.3  1419.8     1419.8  1500900
238  2020-01-10  1427.6  1434.9  1418.3  1429.7     1429.7  1820700
239  2020-01-13  1436.1  1440.5  1426.0  1439.2     1439.2  1652300
240  2020-01-14  1439.0  1441.8  1428.4  1430.9     1430.9  1558900
241  2020-01-15  1430.2  1441.4  1430.2  1439.2     1439.2  1282700
242  2020-01-16  1447.4  1452.0  1440.9  1451.7     1451.7  1173700
243  2020-01-17  1462.9  1481.3  1458.2  1480.4     1480.4  2396200
244  2020-01-21  1479.1  1491.8  1471.2  1484.4     1484.4  2036700
245  2020-01-22  1491.0  1503.2  1484.9  1485.9     1485.9  1610800
246  2020-01-23  1487.6  1495.5  1482.1  1486.7     1486.7  1351200
247  2020-01-24  1493.6  1495.5  1465.2  1466.7     1466.7  1784600
248  2020-01-27  1431.0  1438.1  1421.2  1433.9     1433.9  1755200
249  2020-01-28  1443.0  1456.0  1432.5  1452.6     1452.6  1577400

We see that it is quite a large dataset and that Pandas has automatically assigned numerical row indices to the DataFrame. Pandas also used the labels that appear in the data in the CSV file to assign the column labels.

When dealing with large datasets like this one, it is often useful just to take a look at the first few rows of data instead of the whole dataset. We can take a look at the first 5 rows of data using the <code>.head()</code> method, as shown below

In [119]:
Google_stock.head()

Date    Open    High     Low   Close  Adj Close   Volume
0  2019-01-31  1103.0  1117.3  1095.4  1116.4     1116.4  1538300
1  2019-02-01  1112.4  1125.0  1104.9  1110.8     1110.8  1462200
2  2019-02-04  1112.7  1132.8  1109.0  1132.8     1132.8  2576500
3  2019-02-05  1124.8  1146.8  1117.2  1146.0     1146.0  3552200
4  2019-02-06  1139.6  1147.0  1112.8  1115.2     1115.2  2105600

We can also take a look at the last 5 rows of data by using the <code>.tail()</code> method:

In [120]:
Google_stock.tail()

Date    Open    High     Low   Close  Adj Close   Volume
248  2020-01-27  1431.0  1438.1  1421.2  1433.9     1433.9  1755200
249  2020-01-28  1443.0  1456.0  1432.5  1452.6     1452.6  1577400
250  2020-01-29  1458.8  1465.4  1446.7  1458.6     1458.6  1077700
251  2020-01-30  1440.0  1457.3  1436.4  1455.8     1455.8  1335200
252  2020-01-31  1468.9  1470.1  1428.5  1434.2     1434.2  2296131

We can also optionally use <code>.head(N)</code> or <code>.tail(N)</code> to display the first and last <code>N</code> rows of data, respectively.

Let's do a quick check to see whether we have any NaN values in our dataset. To do this, we will use the <code>.isnull()</code> method followed by the <code>.any()</code> method to check whether any of the columns contain NaN values.

In [122]:
Google_stock.isnull().any()

Date         False
Open         False
High         False
Low          False
Close        False
Adj Close    False
Volume       False
dtype: bool

We see that we have no <code>NaN</code> values.

When dealing with large datasets, it is often useful to get statistical information from them. Pandas provides the <code>.describe()</code> method to get descriptive statistics on each column of the DataFrame. Let's see how this works:

We get descriptive statistics on our stock data

In [124]:
Google_stock.describe()

Open    High     Low   Close  Adj Close   Volume
count   253.0   253.0   253.0   253.0      253.0  2.5e+02
mean   1216.8  1226.7  1208.5  1218.4     1218.4  1.4e+06
std      98.8    98.9    99.3    99.3       99.3  6.2e+05
min    1042.9  1047.5  1025.0  1036.2     1036.2  3.5e+05
25%    1144.0  1150.8  1131.8  1144.9     1144.9  1.1e+06
50%    1197.6  1206.4  1190.1  1200.5     1200.5  1.3e+06
75%    1274.0  1281.6  1260.5  1273.7     1273.7  1.6e+06
max    1493.6  1503.2  1484.9  1486.7     1486.7  6.2e+06

If desired, we can apply the <code>.describe()</code> method on a single column as shown below:

We get descriptive statistics on a single column of our DataFrame

In [127]:
Google_stock['Adj Close'].describe()

count     253.0
mean     1218.4
std        99.3
min      1036.2
25%      1144.9
50%      1200.5
75%      1273.7
max      1486.7
Name: Adj Close, dtype: float64

Similarly, you can also look at one statistic by using one of the many statistical functions Pandas provides. Let's look at some examples:

We print information about our DataFrame

In [128]:
print()
print('Maximum values of each column:\n', Google_stock.max())
print()
print('Minimum Close value:', Google_stock['Close'].min())
print()
print('Average value of each column:\n', Google_stock.mean())


Maximum values of each column:
 Date         2020-01-31
Open              1e+03
High              2e+03
Low               1e+03
Close             1e+03
Adj Close         1e+03
Volume          6207000
dtype: object

Minimum Close value: 1036.22998

Average value of each column:
 Open         1.2e+03
High         1.2e+03
Low          1.2e+03
Close        1.2e+03
Adj Close    1.2e+03
Volume       1.4e+06
dtype: float64


Another important statistical measure is data correlation. Data correlation can tell us, for example, if the data in different columns are correlated. We can use the <code>.corr()</code> method to get the correlation between different columns, as shown below:

We display the correlation between columns

In [130]:
Google_stock.corr()

Open     High      Low    Close  Adj Close   Volume
Open       1.0e+00  1.0e+00  1.0e+00  9.9e-01    9.9e-01 -3.6e-02
High       1.0e+00  1.0e+00  1.0e+00  1.0e+00    1.0e+00 -2.0e-02
Low        1.0e+00  1.0e+00  1.0e+00  1.0e+00    1.0e+00 -5.7e-02
Close      9.9e-01  1.0e+00  1.0e+00  1.0e+00    1.0e+00 -3.6e-02
Adj Close  9.9e-01  1.0e+00  1.0e+00  1.0e+00    1.0e+00 -3.6e-02
Volume    -3.6e-02 -2.0e-02 -5.7e-02 -3.6e-02   -3.6e-02  1.0e+00

A correlation value of 1 tells us there is a high correlation and a correlation of 0 tells us that the data is not correlated at all.